src/txn/txn_timestamp.c

```diff
@@ -740,74 +746,103 @@ __wt_txn_set_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
+	WT_TXN *prev, *txn;
 	WT_TXN_GLOBAL *txn_global;
-	wt_timestamp_t oldest_ts;
-	char ts_string[WT_TS_INT_STRING_SIZE];
+	wt_timestamp_t oldest_ts, timestamp;
+	char ts_string[2][WT_TS_INT_STRING_SIZE];
 
+	txn = &session->txn;
 	txn_global = &S2C(session)->txn_global;
 
 	WT_RET(__wt_config_gets_def(session,
 	    cfg, "prepare_timestamp", 0, &cval));
-	if (cval.len > 0) {
-		if (F_ISSET(&session->txn, WT_TXN_HAS_TS_COMMIT))
-			WT_RET_MSG(session, EINVAL,
-			    "commit timestamp should not have been set before "
-			    "prepare transaction");
+	if (cval.len == 0)
+		WT_RET_MSG(session, EINVAL, "prepare timestamp is required");
 
-		WT_RET(__wt_txn_parse_timestamp(
-		    session, "prepare", timestamp, &cval));
+	if (F_ISSET(&session->txn, WT_TXN_HAS_TS_COMMIT))
+		WT_RET_MSG(session, EINVAL, "commit timestamp "
+		    "should not have been set before prepare transaction");
 
+	WT_RET(__wt_txn_parse_timestamp(session, "prepare", &timestamp, &cval));
+
+	__wt_readlock(session, &txn_global->read_timestamp_rwlock);
+
+	oldest_ts = txn_global->oldest_timestamp;
+
+	/*
+	 * Prepare timestamp must be later/greater than latest active
+	 * read timestamp, if any.
+	 */
+	prev = TAILQ_LAST(
+	    &txn_global->read_timestamph, __wt_txn_rts_qh);
+	while (prev != NULL) {
+		/* Skip self and non-active transactions. */
+		if (prev->clear_read_q || prev == txn) {
+			prev = TAILQ_PREV(
+			    prev, __wt_txn_rts_qh, read_timestampq);
+			continue;
+		}
+
+		if (prev->read_timestamp >= timestamp) {
+			__wt_readunlock(session,
+			    &txn_global->read_timestamp_rwlock);
+			__wt_timestamp_to_string(
+			    prev->read_timestamp, ts_string[0]);
+			WT_RET_MSG(session, EINVAL,
+			    "prepare timestamp %.*s not later than an "
+			    "active read timestamp %s ",
+			    (int)cval.len, cval.str, ts_string[0]);
+		}
+		break;
+	}
+
+	/* Unlock here to have less code branches. */
+	__wt_readunlock(session, &txn_global->read_timestamp_rwlock);
+	/*
+	 * Check whether the prepare timestamp is lesser / earlier to the
+	 * oldest timestamp.
+	 */
+	if (timestamp < oldest_ts) {
 		/*
-		 * Prepare timestamp must be later/greater than latest active
-		 * read timestamp.
+		 * Check whether the prepare timestamp need to be rounded up to
+		 * the oldest timestamp.
 		 */
-		__wt_readlock(session, &txn_global->read_timestamp_rwlock);
-		prev = TAILQ_LAST(&txn_global->read_timestamph,
-		    __wt_txn_rts_qh);
-		while (prev != NULL) {
+		if (F_ISSET(txn, WT_TXN_TS_ROUND_PREPARED)) {
 			/*
-			 * Skip any transactions that are not active.
+			 * Check that there are no active readers - that would
+			 * be a violation of preconditions for rounding
+			 * timestamps of prepared transactions.
 			 */
-			if (prev->clear_read_q) {
-				prev = TAILQ_PREV(
-				    prev, __wt_txn_rts_qh, read_timestampq);
-				continue;
-			}
-			if (prev->read_timestamp >= *timestamp) {
-				__wt_readunlock(session,
-				    &txn_global->read_timestamp_rwlock);
+			/*
+			 * TODO:
```

Comment:
Is there value in leaving a TODO here? If the assertion fires, we will know there is a problem. I thought the logic above was supposed to skip the checkpointing thread?